Sources/SwiftFormatPrettyPrint/TokenStreamCreator.swift

```diff
@@ -2274,14 +2281,46 @@ private final class TokenStreamCreator: SyntaxVisitor {
   /// the expression, keeping the entire expression together when possible, before breaking inside
   /// the expression. This is a hand-crafted list of expressions that generally look better when the
   /// break(s) before the expression fire before breaks inside of the expression.
-  private func maybeGroupAroundSubexpression(_ expr: ExprSyntax) {
+  private func maybeGroupAroundSubexpression(
+    _ expr: ExprSyntax, combiningOperator operatorExpr: ExprSyntax? = nil
+  ) {
     switch expr {
-    case is FunctionCallExprSyntax, is MemberAccessExprSyntax, is SubscriptExprSyntax:
+    case is MemberAccessExprSyntax, is SubscriptExprSyntax:
       before(expr.firstToken, tokens: .open)
       after(expr.lastToken, tokens: .close)
     default:
       break
     }
+
+    // Function calls in expressions are conditionally grouped based on whether they are preceded by
+    // an assigning operator. Assigning the result of a function call is allowed to break the
+    // the rectangle rule, in order to save vertical space.
+    if expr is FunctionCallExprSyntax,
+      let operatorExpr = operatorExpr, !isAssigningOperator(operatorExpr)
+    {
+      before(expr.firstToken, tokens: .open)
+      after(expr.lastToken, tokens: .close)
+    }
+  }
+
+  /// Returns whether the given operator behaves as an assignment, to assign a right-hand-side to a
+  /// left-hand-side in a SequenceExpr.
+  ///
+  /// Assignment is defined as either being an assignment operator (i.e. `=`) or any operator that
+  /// uses "assignment" precedence.
+  private func isAssigningOperator(_ operatorExpr: ExprSyntax) -> Bool {
+    if operatorExpr is AssignmentExprSyntax {
+      return true
+    }
+    if let binaryOperator = operatorExpr as? BinaryOperatorExprSyntax {
+      let operatorText = binaryOperator.operatorToken.text
+      if let precedence = operatorContext.infixOperator(named: operatorText)?.precedenceGroup,
+        precedence === operatorContext.precedenceGroup(named: .assignment)
```

Comment:
Hi guys 👋 

Sorry to bump into this PR out of the blue, but I'm currently adding `swift-format` to our project and this was one of the "problems" we found with the current implementation, so thanks for tackling it!

Would it be reasonable to extend this rule to not just apply to `AssignmentPrecedence` operators, but to any operator with precedence *equal or higher* than `AssignmentPrecedence`?

In our particular case, we use ReactiveSwift which adds the custom operator `<~` (binding operator) defined as such:

```swift
precedencegroup BindingPrecedence {
	associativity: right

	// Binds tighter than assignment but looser than everything else
	higherThan: AssignmentPrecedence
}
```

It's essentially used whenever an assignment is made (hence the precedence relationship), and it manifests the same "wrong" formatting behavior by indenting expression into the next line if it doesn't fit (which happens most of the time since we're dealing with complex reactive chains).

What do you say? Thanks! 🙏

Reply:
Just for context - you can see what operators we currently support [here](https://github.com/apple/swift-format/blob/master/Sources/SwiftFormatPrettyPrint/OperatorContext.swift).

If you're interested, we'd be happy to review a PR that adds support for reading custom operators in from the formatter's configuration!

Reply:
Thanks for the feedback! 🙇 

I'm still **very** unfamiliar with `swift-format`'s codebase, so pardon my ignorance 😅. What you mention makes complete sense. I noticed some types representing operator precedences (e.g. `.assignment`), but didn't dive into the code to see if it was effectively doable or made sense. Guess I was being too optimistic!

However, I do intend to invest some time in `swift-format` as I want to try and make some rules more configurable as well as add support for some new ones, so we can automatically enforce as much of our style guide as possible. This change will definitely be the list of stuff to do, but I think I'll start with some "easier" additions first 😇.

Keep it up!